test/document_stores/test_sql.py

```diff
@@ -62,6 +62,42 @@ def test_sql_write_different_documents_same_vector_id(self, ds):
         with pytest.raises(Exception, match=r"(?i)unique"):
             ds.write_documents([doc2], index="index3")
 
+    @pytest.mark.integration
+    def test_sql_get_documents_using_nested_filters_about_classification(self, ds):
+        documents = [
+            Document(
+                content="That's good. I like it.",
+                id="1",
+                meta={
+                    "classification": {
+                        "label": "LABEL_1",
+                        "score": 0.694,
+                        "details": {"LABEL_1": 0.694, "LABEL_0": 0.306},
+                    }
+                },
+            ),
+            Document(
+                content="That's bad. I don't like it.",
+                id="2",
+                meta={
+                    "classification": {
+                        "label": "LABEL_0",
+                        "score": 0.898,
+                        "details": {"LABEL_0": 0.898, "LABEL_1": 0.102},
+                    }
+                },
+            ),
+        ]
+        ds.write_documents(documents)
+
+        assert ds.get_document_count() == 2
+        assert len(ds.get_all_documents(filters={"classification.score": {"$gt": 0.1}})) == 2
+        assert len(ds.get_all_documents(filters={"classification.label": ["LABEL_1", "LABEL_0"]})) == 2
+        assert len(ds.get_all_documents(filters={"classification.score": {"$gt": 0.8}})) == 1
+        assert len(ds.get_all_documents(filters={"classification.label": ["LABEL_1"]})) == 1
+        assert len(ds.get_all_documents(filters={"classification.score": {"$gt": 0.95}})) == 0
+        assert len(ds.get_all_documents(filters={"classification.label": ["LABEL_100"]})) == 0
+
```

Comment:
Hello guys!

It seems that my idea works but my test is failing :smile:
And I need some help to better understand... 

Only one document is stored in DB instead of 2.
I can reproduce this issue in the tests every time we want to write 2 documents with generic metadata using standard `duplicate_documents="overwrite"`. It happens only in the tests...

In this [notebook](https://colab.research.google.com/drive/1E5KftwcLDIlZbp7LUuiomBsgn1CSNo3x?usp=sharing) I tried to reproduce the test and it works fine. (Also in my local installation).

### Triying to understand...

<details>
  <summary>I also captured the complete SQLAlchemy logs</summary>
INFO     sqlalchemy.engine.Engine:base.py:139 BEGIN (implicit; DBAPI should not BEGIN due to autocommit mode)
INFO     sqlalchemy.engine.Engine:base.py:139 DELETE FROM meta_document WHERE meta_document.document_id = ?
INFO     sqlalchemy.engine.Engine:base.py:139 [generated in 0.00113s] ('1',)
INFO     sqlalchemy.engine.Engine:base.py:139 SELECT document.id AS document_id, document.created_at AS document_created_at, document.updated_at AS document_updated_at, document.content AS document_content, document.content_type AS document_content_type, document."index" AS document_index, document.vector_id AS document_vector_id, meta_document_1.id AS meta_document_1_id, meta_document_1.created_at AS meta_document_1_created_at, meta_document_1.updated_at AS meta_document_1_updated_at, meta_document_1.name AS meta_document_1_name, meta_document_1.value AS meta_document_1_value, meta_document_1.document_id AS meta_document_1_document_id, meta_document_1.document_index AS meta_document_1_document_index 
FROM document LEFT OUTER JOIN meta_document AS meta_document_1 ON document.id = meta_document_1.document_id AND document."index" = meta_document_1.document_index 
WHERE document.id = ? AND document."index" = ?
INFO     sqlalchemy.engine.Engine:base.py:139 [generated in 0.00666s] ('1', 'test.document_stores.test_sql')
DEBUG    sqlalchemy.engine.Engine:base.py:150 Col ('document_id', 'document_created_at', 'document_updated_at', 'document_content', 'document_content_type', 'document_index', 'document_vector_id', 'meta_document_1_id', 'meta_document_1_created_at', 'meta_document_1_updated_at', 'meta_document_1_name', 'meta_document_1_value', 'meta_document_1_document_id', 'meta_document_1_document_index')
INFO     sqlalchemy.engine.Engine:base.py:139 ROLLBACK using DBAPI connection.rollback(), DBAPI should ignore due to autocommit mode
INFO     sqlalchemy.engine.Engine:base.py:139 BEGIN (implicit; DBAPI should not BEGIN due to autocommit mode)
INFO     sqlalchemy.engine.Engine:base.py:139 DELETE FROM meta_document WHERE meta_document.document_id = ?
INFO     sqlalchemy.engine.Engine:base.py:139 [cached since 0.04453s ago] ('2',)
INFO     sqlalchemy.engine.Engine:base.py:139 SELECT document.id AS document_id, document.created_at AS document_created_at, document.updated_at AS document_updated_at, document.content AS document_content, document.content_type AS document_content_type, document."index" AS document_index, document.vector_id AS document_vector_id, meta_document_1.id AS meta_document_1_id, meta_document_1.created_at AS meta_document_1_created_at, meta_document_1.updated_at AS meta_document_1_updated_at, meta_document_1.name AS meta_document_1_name, meta_document_1.value AS meta_document_1_value, meta_document_1.document_id AS meta_document_1_document_id, meta_document_1.document_index AS meta_document_1_document_index 
FROM document LEFT OUTER JOIN meta_document AS meta_document_1 ON document.id = meta_document_1.document_id AND document."index" = meta_document_1.document_index 
WHERE document.id = ? AND document."index" = ?
INFO     sqlalchemy.engine.Engine:base.py:139 [cached since 0.03686s ago] ('2', 'test.document_stores.test_sql')
DEBUG    sqlalchemy.engine.Engine:base.py:150 Col ('document_id', 'document_created_at', 'document_updated_at', 'document_content', 'document_content_type', 'document_index', 'document_vector_id', 'meta_document_1_id', 'meta_document_1_created_at', 'meta_document_1_updated_at', 'meta_document_1_name', 'meta_document_1_value', 'meta_document_1_document_id', 'meta_document_1_document_index')
INFO     sqlalchemy.engine.Engine:base.py:139 INSERT INTO document (id, content, content_type, "index", vector_id) VALUES (?, ?, ?, ?, ?)
INFO     sqlalchemy.engine.Engine:base.py:139 [generated in 0.00072s] ('2', '"That\'s bad. I don\'t like it."', 'text', 'test.document_stores.test_sql', None)
INFO     sqlalchemy.engine.Engine:base.py:139 INSERT INTO meta_document (id, name, value, document_id, document_index) VALUES (?, ?, ?, ?, ?)
INFO     sqlalchemy.engine.Engine:base.py:139 [generated in 0.00095s] ('940f135d-90ce-4380-ac70-ff026fac537d', 'classification.label', 'LABEL_0', '2', 'test.document_stores.test_sql')
INFO     sqlalchemy.engine.Engine:base.py:139 INSERT INTO meta_document (id, name, value, document_id, document_index) VALUES (?, ?, ?, ?, ?)
INFO     sqlalchemy.engine.Engine:base.py:139 [cached since 0.007702s ago] ('d4d985b6-eda2-4a43-b975-c092a5f78e31', 'classification.score', 0.898, '2', 'test.document_stores.test_sql')
INFO     sqlalchemy.engine.Engine:base.py:139 INSERT INTO meta_document (id, name, value, document_id, document_index) VALUES (?, ?, ?, ?, ?)
INFO     sqlalchemy.engine.Engine:base.py:139 [cached since 0.01233s ago] ('e44b66b1-bcc9-43a2-ab7d-10e02deddfc7', 'classification.details', "{'LABEL_0': 0.898, 'LABEL_1': 0.102}", '2', 'test.document_stores.test_sql')
INFO     sqlalchemy.engine.Engine:base.py:139 COMMIT using DBAPI connection.commit(), DBAPI should ignore due to autocommit mode
INFO     sqlalchemy.engine.Engine:base.py:139 BEGIN (implicit; DBAPI should not BEGIN due to autocommit mode)
INFO     sqlalchemy.engine.Engine:base.py:139 SELECT anon_1.document_id AS anon_1_document_id 
FROM (SELECT document.id AS document_id, row_number() OVER (ORDER BY document.id) AS rownum 
FROM document) AS anon_1 
WHERE rownum % 10000=1
INFO     sqlalchemy.engine.Engine:base.py:139 [generated in 0.00065s] ()
DEBUG    sqlalchemy.engine.Engine:base.py:150 Col ('anon_1_document_id',)
DEBUG    sqlalchemy.engine.Engine:base.py:150 Row ('2',)
INFO     sqlalchemy.engine.Engine:base.py:139 SELECT document.id AS document_id, document.content AS document_content, document.content_type AS document_content_type, document.vector_id AS document_vector_id 
FROM document 
WHERE document."index" = ? AND document.id >= ? ORDER BY document.id
INFO     sqlalchemy.engine.Engine:base.py:139 [generated in 0.00067s] ('test.document_stores.test_sql', '2')
DEBUG    sqlalchemy.engine.Engine:base.py:150 Col ('document_id', 'document_content', 'document_content_type', 'document_vector_id')
DEBUG    sqlalchemy.engine.Engine:base.py:150 Row ('2', "That's bad. I don't like it.", 'text', None)
INFO     sqlalchemy.engine.Engine:base.py:139 ROLLBACK using DBAPI connection.rollback(), DBAPI should ignore due to autocommit mode
INFO     sqlalchemy.engine.Engine:base.py:139 BEGIN (implicit; DBAPI should not BEGIN due to autocommit mode)
INFO     sqlalchemy.engine.Engine:base.py:139 SELECT meta_document.document_id AS meta_document_document_id, meta_document.name AS meta_document_name, meta_document.value AS meta_document_value 
FROM meta_document 
WHERE meta_document.document_id IN (?)
INFO     sqlalchemy.engine.Engine:base.py:139 [generated in 0.00134s] ('2',)
DEBUG    sqlalchemy.engine.Engine:base.py:150 Col ('meta_document_document_id', 'meta_document_name', 'meta_document_value')
DEBUG    sqlalchemy.engine.Engine:base.py:150 Row ('2', 'classification.label', 'LABEL_0')
DEBUG    sqlalchemy.engine.Engine:base.py:150 Row ('2', 'classification.score', '0.898')
DEBUG    sqlalchemy.engine.Engine:base.py:150 Row ('2', 'classification.details', "{'LABEL_0': 0.898, 'LABEL_1': 0.102}")
INFO     sqlalchemy.engine.Engine:base.py:139 ROLLBACK using DBAPI connection.rollback(), DBAPI should ignore due to autocommit mode
INFO     sqlalchemy.engine.Engine:base.py:139 BEGIN (implicit; DBAPI should not BEGIN due to autocommit mode)
INFO     sqlalchemy.engine.Engine:base.py:139 SELECT count(*) AS count_1 
FROM (SELECT document.id AS document_id, document.created_at AS document_created_at, document.updated_at AS document_updated_at, document.content AS document_content, document.content_type AS document_content_type, document."index" AS document_index, document.vector_id AS document_vector_id 
FROM document 
WHERE document."index" = ?) AS anon_1
INFO     sqlalchemy.engine.Engine:base.py:139 [generated in 0.00108s] ('test.document_stores.test_sql',)
DEBUG    sqlalchemy.engine.Engine:base.py:150 Col ('count_1',)
DEBUG    sqlalchemy.engine.Engine:base.py:150 Row (1,)
</details>

**From the logs, you can see that for document 1 possible old metadata are deleted. Then the document is not inserted. Why?
For Doc. 2, old metadata are deleted, then it is inserted.**

I also tried to play with `isolation_level` in DS initialization and with `synchronize_session` in  https://github.com/deepset-ai/haystack/blob/3319ef6d1c8f0b8a4d80d0f531c6ea0fd1c02e6e/haystack/document_stores/sql.py#L420
but nothing solved the problem.

### Workaround
```
ds.write_documents([documents[0]])
ds.write_documents([documents[1]])
```

If we perform two operations, everything works fine.

@bogdankostic @masci @ZanSara 
do you have any suggestions or ideas?

Reply:
@anakin87 I can reproduce locally on my Mac (`assert 1 == 2`), let me investigate and I'll get back to you with my findings

Reply:
@anakin87 the problem is here: https://github.com/deepset-ai/haystack/blob/main/test/conftest.py#L131

When we call `self._flatten_classification_meta_fields` the session is rolled back ¯\_(ツ)_/¯. This also explains why this works in colab (and in general when you don't run that code through pytest).

@ZanSara do you know if we can safely remove that evil piece of code in `conftest.py`?

Reply:
It was introduced in https://github.com/deepset-ai/haystack/pull/761

Reply:
If we agree that this hack should be removed, I can take care of it in this PR or another...

Reply:
One more thing: every time we tweak the SQL tests we should try to run them on PostgreSQL too. They rot very fast on DBs other than SQLite

Reply:
> By the way, before we move on: are we sure that this failure is not actually correct, so that there are no uncommitted transactions somewhere in the new diff?

Not sure but the check would be wrong anyways, see here https://github.com/deepset-ai/haystack/blob/main/test/conftest.py#L121-L122
```
    method = object.__getattribute__(self, attr)
    if callable(method):
        try:
            self.session.rollback()
```
this is triggered by just _accessing_ `self._flatten_classification_meta_fields` (since it's a callable)

Reply:
> One more thing: every time we tweak the SQL tests we should try to run them on PostgreSQL too. They rot very fast on DBs other than SQLite
now that they're better isolated we should just run Postgres in the CI (also possible side by side with SQLite)

Reply:
> > By the way, before we move on: are we sure that this failure is not actually correct, so that there are no uncommitted transactions somewhere in the new diff?
> 
> Not sure but the check would be wrong anyways, see here https://github.com/deepset-ai/haystack/blob/main/test/conftest.py#L121-L122
> 
> ```
>     method = object.__getattribute__(self, attr)
>     if callable(method):
>         try:
>             self.session.rollback()
> ```
> 
> this is triggered by just _accessing_ `self._flatten_classification_meta_fields` (since it's a callable)

Yes, but in theory the rollback should have no effect if everything is committed before that call... right? However I won't argue, this is generally bad and I want it gone too :smile:

Reply:
- `self._flatten_classification_meta_fields` seems safe: it only modifies the `meta` dictionary and not ORM objects. Furthermore, I tested this change locally without `AUTOCOMMIT`

- I tried to remove the hack in conftest and locally run the tests for `SQLDocumentStore` using SQLite and PostgreSQL. They are ok.

- On the other hand, using PostgreSQL, the asserts involving `$gt` operator in  `test_sql_get_documents_using_nested_filters_about_classification` fail (string vs float comparison is not automatically performed in PostgreSQL). Other asserts in this test are OK (eg. filtering for a specific label or for a list of labels).
So now that #3547 is in progress, the current PR doesn't seem to add much: it only offers the possibility of filtering for a specific label or for a list of labels; it is also very specific to classification metadata, while the SQL DS does not fully support filtering.

Based on those considerations, what about closing this PR and opening another one just to remove that hack in conftest? (or we can also remove the hack in this same PR.)

Reply:
@ZanSara @masci please let me know how to proceed 😉

Reply:
@anakin87 let's remove the hack, up to you if you want to "repurpose" this PR or opening a new one.

I wouldn't spend time on the Postgres problem, after refactoring all the tests I'd like to explore the opportunity to run SQL tests in the CI against postgres, at that point we'll fix what's broken.